docs/madr/decisions/049-meshexternalservice-api.md

```diff
@@ -0,0 +1,474 @@
+ destination:
+   type: Passthrough
+```
+
+In this case, for port `80`, we require HTTP/2 configuration in the cluster settings. This necessitates having a separate configuration for each cluster.
+
+clusterName: `{policyName}_{port}`, so for the above policy name would looks like: `httpbin_80` and `httpbin_443`
+
+#### Extensability
+
+Provided model allows creating separate plugins which can integrate with different cloud providers or components. The user can create it's own custom configuration and later in the code create XDS configuration based on it. 
+
+Example:
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: example
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+spec:
+  match:
+  - type: InternalVIP
+    value: example.ext.svc.local
+    port: 80
+    protocol: http
+ destination:
+   type: Extension
+   extension:
+     type: Lambda 
+     config:
+       arn: arn:aws:lambda:us-west-2:123456789012:function:my-function
+```
+
+#### IP intersections 
+
+It's possible for users to create a policy where the IP ranges intersect with each other. We cannot validate this during policy creation, but during configuration generation, we are able to check if IPs intersect. In such cases, we can log a message stating:
+
+> external-service1 and external-service2 have overlapping IPs X.X.X.X, which can disrupt your traffic.
+
+#### Domain generation
+
+The `ExternalService` resource automatically allocated an internal IP each real domain (e.g. `httpbin.com`). While this approach offered benefits such as avoiding the allocation of a listener for a specific port listening on `0.0.0.0`, it obscured the existing domain behind a custom IP. With the introduction of `MeshExternalService`, we expose to users a type called `InternalVIP`, enabling the creation of domains for specific external services.
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: mongo
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+spec:
+  match:
+  - type: InternalVIP # Kuma will generate a domain
+    value: mongo.ext.svc.local
+    port: 27017
+    protocol: tcp
+ destination:
+   type: Regular # Regular|Passthrough|Extension
+   endpoints:
+   - address: 10.0.0.1
+     port: 27017
+   - address: 10.0.0.2
+     port: 27017
+status: 
+  vips:
+  - ip: 242.0.0.1
+    type: Kuma
+    hostname: mongo.ext.svc.local
+```
+
+In this scenario, we generate a custom domain `mongo.ext.svc.local` and allocate the address `242.0.0.1` to it.
+For the purpose of `MeshExternalServices` we are going to take a completely new CIDR by default `242.0.0.0/8`. This CIDR can be configured by the user.
+
+#### Wildcard domain
+
+Wildcard domains are supported only when `destination.type: Passthrough`, which is a default one.
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: kafka
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+spec:
+  match: 
+  - type: Domain
+    value: *.eu-west-3.aws.cloud 
```

Comment:
we could mention that we will support partial `*` like `*-gregw.eu-west-3.aws.cloud` using Matching API

Reply:
matching api? I think simple filterchainmatch with server_name works as well
```
For criteria that allow ranges or wildcards, the most specific value in any of the configured filter chains that matches the incoming connection is going to be used (e.g. for SNI www.example.com the most specific match would be www.example.com, then *.example.com, then *.com, then any filter chain without server_names requirements).
```

Reply:
added note

Reply:
I think there's a thing with partial matches (matches that don't start with a subdomain separator like `.`) no?

Reply:
> matching api? I think simple filterchainmatch with server_name works as well

https://www.envoyproxy.io/docs/envoy/latest/xds/type/matcher/v3/domain.proto#xds-type-matcher-v3-servernamematcher-domainmatcher  

![image](https://github.com/kumahq/kuma/assets/2753650/e4827527-2d2b-4fe5-a52a-98a7103ef423)

Reply:
```
                        single_predicate:
                          input:
                            name: tls
                            typed_config:
                              "@type": type.googleapis.com/envoy.extensions.matching.common_inputs.network.v3.ServerNameInput
                          value_match:
                            safe_regex:
                              google_re2: {}
                              regex: ".*w.httpbin.com"```